spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,197 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.hadoop.HiddenPathFilter;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.Tasks;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.util.SerializableConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import parquet.Preconditions;
+
+public class RemoveOrphanFilesAction implements Action<RemoveOrphanFilesActionResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoveOrphanFilesAction.class);
+
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final SerializableConfiguration hadoopConf;
+  private final FileIO fileIO;
+  private final int partitionDiscoveryParallelism;
+  private final String dataLocation;
+
+  private String allDataFilesTable = null;
+  private Long olderThanTimestamp = null;
+  private boolean dryRun = false;
+
+  RemoveOrphanFilesAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.hadoopConf = new SerializableConfiguration(spark.sessionState().newHadoopConf());
+    this.fileIO = table.io();
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.dataLocation = table.locationProvider().dataLocation();
+  }
+
+  public RemoveOrphanFilesAction allDataFilesTable(String newAllDataFilesTable) {
```

Comment:
This seems awkward, but I'm not sure a better way to do it.

Reply:
Any other ideas are more than welcome :)

Reply:
In other places, we detect path tables by checking `contains("/")`. We could do that here to construct the metadata table names:

```
public String metadataTableName(Table table, String metaTable) {
  String tableName = table.toString()
  if (tableName.contains("/")) {
    return tableName + "#" + metaTable;
  } else {
    return tableName + "." + metaTable;
  }
}
```

Reply:
This almost works. Unfortunately, `table.toString()` might return anything. For example, it will prepend `hive.` for friendly names in the Hive catalog and we won't be able to resolve `hive.db.table.all_data_files`. Exposing a correct `TableIdentifier` in `Table` would help but that would mean modifying public `BaseTable` as well.

Reply:
Okay, I thought this would work because it is what we do in our Spark 2.4 build. Since we are using DSv2 catalogs, when the catalog adds its name to the identifier we actually get a working multi-catalog identifier.

Maybe we should add something to remove `hive.` for now, and take it out for Spark 3.0.